codecs/jxl/dec/jxl_dec.cpp

```diff
@@ -1,31 +1,77 @@
+  JxlDecoder* dec = JxlDecoderCreate(nullptr);
+  EXPECT_EQ(JXL_DEC_SUCCESS,
+            JxlDecoderSubscribeEvents(
+                dec, JXL_DEC_BASIC_INFO | JXL_DEC_COLOR_ENCODING | JXL_DEC_FULL_IMAGE));
   auto next_in = (const uint8_t*)data.c_str();
   auto avail_in = data.size();
-  EXPECT_EQ(JPEGXL_DEC_BASIC_INFO, JpegxlDecoderProcessInput(dec, &next_in, &avail_in));
+  EXPECT_EQ(JXL_DEC_BASIC_INFO, JxlDecoderProcessInput(dec, &next_in, &avail_in));
   size_t buffer_size;
-  const JpegxlPixelFormat format = {4, JPEGXL_LITTLE_ENDIAN, JPEGXL_TYPE_UINT8};
-  EXPECT_EQ(JPEGXL_DEC_SUCCESS, JpegxlDecoderImageOutBufferSize(dec, &format, &buffer_size));
-  JpegxlBasicInfo info;
-  EXPECT_EQ(JPEGXL_DEC_SUCCESS, JpegxlDecoderGetBasicInfo(dec, &info));
+  const JxlPixelFormat format = {4, JXL_LITTLE_ENDIAN, JXL_TYPE_UINT8};
+  EXPECT_EQ(JXL_DEC_SUCCESS, JxlDecoderImageOutBufferSize(dec, &format, &buffer_size));
+  JxlBasicInfo info;
+  EXPECT_EQ(JXL_DEC_SUCCESS, JxlDecoderGetBasicInfo(dec, &info));
+
+  EXPECT_EQ(JXL_DEC_COLOR_ENCODING, JxlDecoderProcessInput(dec, &next_in, &avail_in));
+  size_t icc_size;
+  EXPECT_EQ(JXL_DEC_SUCCESS,
+            JxlDecoderGetICCProfileSize(dec, JXL_COLOR_PROFILE_TARGET_DATA, &icc_size));
+  std::vector<uint8_t> icc_profile(icc_size);
+  EXPECT_EQ(JXL_DEC_SUCCESS,
+            JxlDecoderGetColorAsICCProfile(dec, JXL_COLOR_PROFILE_TARGET_DATA, icc_profile.data(),
+                                           icc_profile.size()));
+
   std::vector<uint8_t> pixels(buffer_size);
-  EXPECT_EQ(JPEGXL_DEC_SUCCESS, JpegxlDecoderSetImageOutBuffer(dec, &format, pixels.data(), pixels.size()));
-  EXPECT_EQ(JPEGXL_DEC_FULL_IMAGE, JpegxlDecoderProcessInput(dec, &next_in, &avail_in));
-  JpegxlDecoderDestroy(dec);
-  return ImageData.new_(Uint8ClampedArray.new_(typed_memory_view(pixels.size(), pixels.data())), info.xsize, info.ysize);
+  EXPECT_EQ(JXL_DEC_SUCCESS,
+            JxlDecoderSetImageOutBuffer(dec, &format, pixels.data(), pixels.size()));
+  EXPECT_EQ(JXL_DEC_FULL_IMAGE, JxlDecoderProcessInput(dec, &next_in, &avail_in));
+  JxlDecoderDestroy(dec);
+#undef EXPECT_EQ
+
+#define EXPECT_TRUE(a)  \
+  if (!(a)) {           \
+    return val::null(); \
+  }
+
+  // Convert to sRGB.
+  skcms_ICCProfile jxl_profile;
+  // If the image is encoded in its original color space, the decoded data will be in the color
+  // space defined by the decoded ICC profile. Otherwise, it is in Linear sRGB. TODO: the decoded
+  // color profile should also be Linear sRGB if !uses_original_profile.
+  if (info.uses_original_profile) {
+    EXPECT_TRUE(skcms_Parse(icc_profile.data(), icc_profile.size(), &jxl_profile));
+  } else {
+    auto s = jxl::ColorEncoding::LinearSRGB(/*gray=*/false);
+    EXPECT_TRUE(s.CreateICC());
+    EXPECT_TRUE(skcms_Parse(s.ICC().data(), s.ICC().size(), &jxl_profile));
+  }
+  EXPECT_TRUE(skcms_Transform(
```

Comment:
Just to double-check - is it guaranteed to be safe to use same source & dest for pixel data?

Reply:
Yes, the skcms doc explicitly states it's OK as long as it's the same pixel format.

Reply:
I see you changed this in the last version - was this not ok after all?

Reply:
As, we **don't** use the same format (anymore?), so makes sense to change it.

Reply:
Yes, indeed with intermediate floats we can't use the same buffer.